### PR TITLE
Tbl087/adding more secure message endpoints

### PR DESCRIPTION
**Pull request by andrewmil**
Adding the remaining API endpoints required for the secure-messaging view in ras-frontstage

- Now endpoints at `/messages_list`, `/message` and `/send_message`
- Errors are now returned from the API with code's of the form `FAXXX`
- Calls are now made to the case and party services so additional config needed for these
- Request handler has been simplified to only catch `ConnectionError` and return an error message with code
- Descriptions of error codes can be found at https://github.com/ONSdigital/ras-frontstage-api/wiki/Errors

This PR works in conjunction with frontstage changes at https://github.com/ONSdigital/ras-frontstage/pull/227

### CODE REVIEW

**Comment by benjefferies** [comment]
I was confused to why we were using flask-restplus over standard flask but I had a look into the documentation and it looks really nice. I specifically like the swagger integration https://flask-restplus.readthedocs.io/en/stable/swagger.html it has and think we should be using this for our apis.

File: frontstage_api/Resources/secure_messaging.py [hunk 1]
```diff
@@ -17,10 +17,105 @@ class GetMessagesList(Resource):
...
+
+        # If the form was submitted with errors and is part of an existing thread, return the last message from thread
+        if message.get('error', {}).get('code') == 'FA006' and message_json.get('thread_id'):
+            thread_message = secure_messaging_controllers.get_thread_message(encoded_jwt, message_json['thread_id'], party_id)
+            message['error']['data'] = {**message['error']['data'], "thread_message": thread_message}
+
+        return make_response(jsonify(message), 200)
 
 
 api.add_resource(GetMessagesList, '/messages_list')
```
Comment (benjefferies) [inline frontstage_api/Resources/secure_messaging.py]: It would be good if we could add the routes on the class. It makes it quite hard to follow if it's at the bottom

File: frontstage_api/Resources/secure_messaging.py [hunk 2]
```diff
@@ -17,10 +17,105 @@ class GetMessagesList(Resource):
...
+        # If the messages were returned with errors return the error message
+        if messages.get('error'):
+            return make_response(jsonify(messages), 200)
+
+        messages["unread_messages_total"] = secure_messaging_controllers.get_unread_message_total(encoded_jwt)
+
+        return make_response(jsonify(messages), 200)
+
+
+class GetMessageView(Resource):
```
Comment (benjefferies) [inline frontstage_api/Resources/secure_messaging.py]: Would it improve the readability if we split the resources out into separate files?

File: frontstage_api/decorators/api_error_handler.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from frontstage_api.exceptions.exceptions import ApiError
+
+
+def api_error_handler(f):
```
Comment (benjefferies) [inline frontstage_api/decorators/api_error_handler.py]: Why can't we use the flask api.errorhandler here?

**Comment by andrewmil** [comment]
I'm a fan of that swagger documentation it provides as well. It would be very easy to implement/support, would be a lot faster/reliable than manually updating an API.md for every update we do.

File: frontstage_api/common/request_handler.py [hunk 4]
```diff
@@ -1,43 +1,28 @@
...
-    except ConnectTimeout as e:
-        logger.error('Connection to remote server timed out', method=method, url=url, exception=str(e))
-        if fail:
-            raise FailedRequest(method, url, e)
-        else:
-            return {
-                "error": "Connection timeout",
-                "url": url
-            }
     except ConnectionError as e:
```
Comment (benjefferies) [inline frontstage_api/common/request_handler.py]: Do we need to catch the ConnectionError and wrap it in an ApiError or can we just catch with @api.errorhandler(ConnectionError)

File: frontstage_api/controllers/case_controller.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+@api_error_handler
+def get_case_by_party_id(party_id):
+    logger.debug('Retrieving case', party_id=party_id)
+    url = app.config['RM_CASE_GET_BY_PARTY'].format(party_id)
+    response = request_handler('GET', url, auth=app.config['BASIC_AUTH'])
+
+    if response.status_code == 204:
+        logger.warning('No case found for party id', party_id=party_id)
+        raise ApiError('FA200')
```
Comment (benjefferies) [inline frontstage_api/controllers/case_controller.py]: We should avoid using arbitrary codes

File: frontstage_api/Resources/secure_messaging.py [hunk 6]
```diff
@@ -5,22 +5,115 @@
 from structlog import wrap_logger
```
Comment (benjefferies) [inline frontstage_api/Resources/secure_messaging.py]: Minor point but we should lowercase the Resources directory for consistency

File: frontstage_api/common/request_handler.py (hunk 4 above)
Comment (andrewmil) [reply frontstage_api/common/request_handler.py]: I was trying to return all errors in the same way from the app (generating them with ApiError). If we're just catching ConnectionError then we lose any reference to the error code associated with it, (assuming that we're happy with the error code method).

File: frontstage_api/Resources/secure_messaging.py (hunk 1 above)
Comment (andrewmil) [reply frontstage_api/Resources/secure_messaging.py]: done. I prefer that way too 👍

File: frontstage_api/controllers/secure_messaging_controllers.py [hunk 7]
```diff
@@ -1,53 +1,144 @@
...
     if response.status_code != 200:
-        logger.error('Error retrieving the unread messages total', status_code=response.status_code)
-        return {
-            "error": "Unexpected status code returned",
-            "url": url
-        }
+        raise ApiError('FA002')
 
     logger.debug('Successfully retrieved the unread message total')
     return json.loads(response.text)
```
Comment (tej99) [inline frontstage_api/controllers/secure_messaging_controllers.py]: Maybe return the extracted number here (.get('total')). So the returned response is {unread_messages_total: xx} instead of {unread_messages_total: {total: xx}}.

File: frontstage_api/controllers/secure_messaging_controllers.py [hunk 8]
```diff
@@ -1,53 +1,144 @@
...
+@api_error_handler
+def get_message(encoded_jwt, message_id, label):
+    logger.debug('Attempting to retrieve message', message_id=message_id, label=label)
+    url = app.config['DRAFT_URL'] if label == 'DRAFT' else app.config['MESSAGE_URL']
+    url = '{}/{}'.format(url, message_id)
+    headers = {"Authorization": encoded_jwt}
+    response = request_handler('GET', url, headers=headers, error_code='FA003')
+
+    if response.status_code != 200:
+        logger.error('Error retrieving the messages', status_code=response.status_code, message_id=message_id, label=label)
```
Comment (tej99) [inline frontstage_api/controllers/secure_messaging_controllers.py]: logging response text might be helpful. Also I don't think this should be logged at error level as SM can also send 404 response and 403 response here.
Maybe the 404 and 403 responses should be handled differently esp if there is a requirement to display different error messages on the ui?

File: frontstage_api/controllers/secure_messaging_controllers.py [hunk 9]
```diff
@@ -1,53 +1,144 @@
...
+@api_error_handler
+def get_thread_message(encoded_jwt, thread_id, party_id):
+    logger.debug('Attempting to retrieve thread message', thread_id=thread_id, party_id=party_id)
+    method = 'GET'
+    url = '{}/{}'.format(app.config['THREAD_URL'], thread_id)
+    headers = {"Authorization": encoded_jwt}
+    response = request_handler(method, url, headers=headers, error_code='FA004')
+
+    if response.status_code != 200:
+        logger.error('Error retrieving the thread message', status_code=response.status_code, thread_id=thread_id, party_id=party_id)
```
Comment (tej99) [inline frontstage_api/controllers/secure_messaging_controllers.py]: same as get_message comment above ^^

File: frontstage_api/controllers/case_controller.py (hunk 5 above)
Comment (andrewmil) [reply frontstage_api/controllers/case_controller.py]: This was a pattern that was discussed with @pricem14pc and @rjweeks. We should probably review it here to make sure it's what we want before adding more to the API though

File: frontstage_api/decorators/api_error_handler.py (hunk 3 above)
Comment (andrewmil) [reply frontstage_api/decorators/api_error_handler.py]: It's a slightly odd scenario, we don't always want to just return the error json when we encounter an error, e.g. for the `/messages_list` if the unread message total fails we still want to return the message list. 
If we use the `api.errorhandler` method it returns the response from that as the overall response to the incoming request.
This decorator just returns the error json as the return from the function it wraps

File: frontstage_api/Resources/secure_messaging.py (hunk 6 above)
Comment (andrewmil) [reply frontstage_api/Resources/secure_messaging.py]: done :+1:

File: frontstage_api/Resources/secure_messaging.py (hunk 2 above)
Comment (andrewmil) [reply frontstage_api/Resources/secure_messaging.py]: I quite like this idea, would keep things a lot tidier

File: frontstage_api/controllers/secure_messaging_controllers.py (hunk 7 above)
Comment (andrewmil) [reply frontstage_api/controllers/secure_messaging_controllers.py]: good idea I'll add this

File: frontstage_api/controllers/secure_messaging_controllers.py (hunk 8 above)
Comment (andrewmil) [reply frontstage_api/controllers/secure_messaging_controllers.py]: I'd argue a 404 or 403 coming at this point should be an error. 
For 404 any message id that is requested from frontstage will have been generated from the `/messages_list` endpoint. So if we're asking for a msg_id that isn't found something has clearly gone wrong.
Similiarly for the 403 if we have a valid token for the frontstage something must have gone wrong for it to be failing when it reaches secure messaging

File: frontstage_api/decorators/api_error_handler.py (hunk 3 above)
Comment (benjefferies) [reply frontstage_api/decorators/api_error_handler.py]: We can still do these things in the api.errorhandler right?
